src/runtime/storageNG/direct-store-muxer.ts

```diff
@@ -41,49 +43,84 @@ export class DirectStoreMuxer<S extends Identified, C extends Identified, T exte
 
   on(callback: ProxyCallback<T>): number {
     this.callbacks.set(this.nextCallbackId, callback);
+    this.callbackIdToMuxIdMap.set(this.nextCallbackId, new Set());
     return this.nextCallbackId++;
   }
 
   off(callback: number): void {
     this.callbacks.delete(callback);
+    for (const muxId of this.callbackIdToMuxIdMap[callback]) {
+      const storeRecord = this.stores[muxId];
+      if (storeRecord.type === 'record') {
+        storeRecord.store.off(storeRecord.idMap.getR(callback));
+        storeRecord.idMap.deleteL(callback);
+      }
+    }
   }
 
-  getLocalModel(muxId: string): CRDTModel<T> {
+  getLocalModel(muxId: string, id: number): CRDTModel<T> {
     const store = this.stores[muxId];
 
     if (store == null) {
-      this.stores[muxId] = {type: 'pending', promise: this.setupStore(muxId)};
+      this.stores[muxId] = {type: 'pending', promise: this.setupStore(muxId, id)};
       return null;
     }
     if (store.type === 'pending') {
       return null;
     } else {
+      if (!store.idMap.hasL(id)) {
+        store.idMap.set(id, this.createListenerForStore(store.store, muxId, id));
+        this.callbackIdToMuxIdMap.get(id).add(muxId);
+      }
       return store.store.localModel;
     }
   }
 
-  private async setupStore(muxId: string): Promise<{type: 'record', store: DirectStore<T>, id: number}> {
+  private async setupStore(muxId: string, callbackId: number): Promise<{type: 'record', store: DirectStore<T>, idMap: BiMap<number, number|Promise<number>>}> {
     const store = await DirectStore.construct<T>({...this.options, mode: StorageMode.Direct, storageKey: this.storageKey.childKeyForBackingElement(muxId)});
-    const record: StoreRecord<T> = {store, id: 0, type: 'record'};
+    const record: StoreRecord<T> = {store, idMap: new BiMap<number, number>(), type: 'record'};
     this.stores[muxId] = record;
-    // Calling store.on may trigger an event; this will be delivered (via processStoreCallback) upstream and may in
-    // turn trigger a request for the localModel. It's important that there's a recorded store in place for the local
-    // model to be retrieved from, even though we don't have the correct id until store.on returns.
-    record.id = store.on(msg => this.processStoreCallback(muxId, msg));
+
+    const storeCallbackId = await this.createListenerForStore(store, muxId, callbackId);
+    record.idMap.set(callbackId, storeCallbackId);
```

Comment:
does the comment that was here no longer apply? Looks like it's still relevant in that this.stores needs to be set before createListenerForStore is called?